pkg/ipcache/ipcache.go

```diff
@@ -0,0 +1,210 @@
+		identityToIPCache:  map[identity.NumericIdentity]map[string]struct{}{},
+		xdsResourceMutator: resourceMutator,
+	}
+}
+
+// Upsert adds / updates  the provided IP and identity into both caches contained
+// within ipc.
+func (ipc *IPCache) Upsert(endpointIP string, identity identity.NumericIdentity) {
+	ipc.mutex.Lock()
+	defer ipc.mutex.Unlock()
+
+	// Update both maps.
+	ipc.ipToIdentityCache[endpointIP] = identity
+
+	_, found := ipc.identityToIPCache[identity]
+	if !found {
+		ipc.identityToIPCache[identity] = map[string]struct{}{}
+	}
+	ipc.identityToIPCache[identity][endpointIP] = struct{}{}
+	endpointIPs := ipc.identityToIPCache[identity]
+
+	// Sort IPs for Upsert into xdsResourceMutator.
+	ipStrings := make([]string, 0, len(endpointIPs))
+	for endpointIP := range endpointIPs {
+		ipStrings = append(ipStrings, endpointIP)
+	}
+	sort.Strings(ipStrings)
+	ipc.xdsResourceMutator.Upsert(envoy.NetworkPolicyHostsTypeURL, identity.StringID(), &envoyAPI.NetworkPolicyHosts{Policy: uint64(identity), HostAddresses: ipStrings}, false)
+
+}
+
+// Delete removes the provided IP-to-security-identity mapping from both caches
+// within ipc.
+func (ipc *IPCache) Delete(endpointIP string) {
+	ipc.mutex.Lock()
+	defer ipc.mutex.Unlock()
+	identity, found := ipc.ipToIdentityCache[endpointIP]
+	if found {
+		delete(ipc.ipToIdentityCache, endpointIP)
+		delete(ipc.identityToIPCache[identity], endpointIP)
+		if len(ipc.identityToIPCache[identity]) == 0 {
+			delete(ipc.identityToIPCache, identity)
+		}
+		ipc.xdsResourceMutator.Delete(envoy.NetworkPolicyHostsTypeURL, identity.StringID(), false)
+	}
+}
+
+// LookupByIP returns the corresponding security identity that endpoint IP maps
+// to within the provided IPCache, as well as if the corresponding entry exists
+// in the IPCache.
+func (ipc *IPCache) LookupByIP(endpointIP string) (identity.NumericIdentity, bool) {
+	ipc.mutex.RLock()
+	identity, exists := ipc.ipToIdentityCache[endpointIP]
+	ipc.mutex.RUnlock()
+	return identity, exists
+}
+
+// LookupByIdentity returns the set of endpoint IPs that have security identity
+// ID, as well as if the corresponding entry exists in the IPCache.
+func (ipc *IPCache) LookupByIdentity(id identity.NumericIdentity) (map[string]struct{}, bool) {
+	ipc.mutex.RLock()
+	ips, exists := ipc.identityToIPCache[id]
+	ipc.mutex.RUnlock()
+	return ips, exists
+}
+
+// IPIdentityMappingOwner is the interface the owner of an identity allocator
+// must implement
+type IPIdentityMappingOwner interface {
+	// TriggerPolicyUpdates will be called whenever a policy recalculation
+	// must be triggered
+	TriggerPolicyUpdates(force bool) *sync.WaitGroup
+}
+
+// GetIPIdentityMapModel returns all known endpoint IP to security identity mappings
+// stored in the key-value store.
+func GetIPIdentityMapModel() {
+	// TODO (ianvernon) return model of ip to identity mapping. For use in CLI.
+	// see GH-2555
+}
+
+func ipIdentityWatcher(owner IPIdentityMappingOwner) {
```

Comment:
Was there a reason not to use the global owner directly? I'm torn on introducing another interface that is, basically, a callback. There isn't too much difference between a closure and an interface in this usage, I guess, apart from needing a type and object instance to satisfy an interface (which happens to work in this case because IPIdentityMappingOwner is a subset of the other owner)

Reply:
Separately, could you add a comment explaining what this watcher is supposed to do?

Reply:
Separately, separately, this watcher seems like it should be a member of the IPCache type? It doesn't interact with much else (I suppose this is related to my `IPIdentitiesPath` above)

Reply:
I did this as it mirrored the behavior in `pkg/identity`, and because it's easier to only specify what functions the owner has to implement (in this case just one). Plus, we don't have to import the Owner from the endpoint package.